### PR TITLE
[Feature] Add Api service optional whether to register ZK.

### PR DESCRIPTION
**Pull request by WangJPLeo**
## Purpose of the pull request

Add Api service optional whether to register ZK.

## Brief change log

related #11263

## Verify this pull request

Manually verified the change by testing locally.

### CODE REVIEW

File: dolphinscheduler-api/pom.xml [hunk 1]
```diff
@@ -64,6 +64,10 @@
             <groupId>org.apache.dolphinscheduler</groupId>
             <artifactId>dolphinscheduler-registry-all</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.dolphinscheduler</groupId>
+            <artifactId>dolphinscheduler-server</artifactId>
+        </dependency>
```
Comment (ruanwenjun) [inline dolphinscheduler-api/pom.xml]: It's better don't add `dolphinscheduler-server` in api, this module might be deprecated.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.validation.Errors;
+import org.springframework.validation.Validator;
+import org.springframework.validation.annotation.Validated;
+
+import lombok.Data;
+
+@Data
+@Validated
+@Configuration
+@ConfigurationProperties(prefix = "api")
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java]:
```suggestion
@ConfigurationProperties(prefix = "register")
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+public class ApiConfig implements Validator {
+
+    private boolean registryEnabled = false;
+    private int listenPort = 7890;
+    private int execThreads = 10;
+    private Duration heartbeatInterval = Duration.ofSeconds(10);
+    private int heartbeatErrorThreshold = 5;
+    private int maxCpuLoadAvg = -1;
+    private double reservedMemory = 0.3;
+    private String apiAddress;
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java]:
```suggestion
    private Duration heartbeatInterval = Duration.ofSeconds(10);
    private int heartbeatErrorThreshold = 5;
    private String apiAddress;
```
The API server doesn't has these properties.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryClient.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import javax.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.Sets;
+
+@Service
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryClient.java:48]: Can we add @ContidionalOnProperty on this bean? there is no need to create thread pool if we don't open the switch.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryServer.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.dolphinscheduler.common.thread.ThreadUtils;
+
+import javax.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryServer.java:33]: Add @ConditionOnProperty?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryClient.java (hunk 4 above)
Comment (WangJPLeo) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryClient.java:48]: thx, it`s a good idea.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryServer.java (hunk 5 above)
Comment (WangJPLeo) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/registry/ApiRegistryServer.java:33]: Yeah, same as ApiRegistryClient.

File: dolphinscheduler-api/pom.xml (hunk 1 above)
Comment (WangJPLeo) [reply dolphinscheduler-api/pom.xml]: At present, HeartBeatTask is in this module. If this module is deprecated, can it be placed under `org/apache/dolphinscheduler/common/thread`?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java (hunk 2 above)
Comment (WangJPLeo) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java]: it`s ok.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java (hunk 3 above)
Comment (WangJPLeo) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/ApiConfig.java]: ok, i'll fix it.

File: dolphinscheduler-api/pom.xml (hunk 1 above)
Comment (WangJPLeo) [reply dolphinscheduler-api/pom.xml]: HeartBeatTask depends on the RegistryClient of the service module, it is not appropriate to move HeartBeatTask to common only. HeartBeatTask and RegistryClient can make a placed PR in the future. @ruanwenjun

Comment (ruanwenjun) [reply dolphinscheduler-api/pom.xml]: There is a pr split `HeartBeatTask ` into MasterHeartBeatTask and `WorkerHeartBeatTask`. In fact, this task is just a schedule task, you don't need to rely on this.

Comment (WangJPLeo) [reply dolphinscheduler-api/pom.xml]: Understood, LGTM.

Comment (WangJPLeo) [reply dolphinscheduler-api/pom.xml]: After reading the PR of #10506, the dependency of Api on the Server service was removed in the latest commit, and the ApiHeartBeatTask was separated to prepare for the next refactoring merge. @ruanwenjun

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/NodeType.java [hunk 6]
```diff
@@ -18,5 +18,5 @@
 package org.apache.dolphinscheduler.common.enums;
 
 public enum NodeType {
-    MASTER, WORKER, DEAD_SERVER
+    API, MASTER, WORKER, DEAD_SERVER
```
Comment (lenboo) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/NodeType.java]: It's better to put the new enum in the last index.

Comment (WangJPLeo) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/NodeType.java]:
```suggestion
    MASTER, WORKER, DEAD_SERVER,API
```

Comment (WangJPLeo) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/enums/NodeType.java]: ok

**Review by lenboo — DISMISSED** [review]
+1

**Comment by WangJPLeo** [comment]
Consider service locking.

**Pull request closed without merging**